lib/bundler/injector.rb

```diff
@@ -90,5 +115,57 @@ def append_to(gemfile_path, new_gem_lines)
         f.puts new_gem_lines
       end
     end
+
+    def remove_gems_from_gemfile(removed_deps, gemfile_path)
+      # store patterns of all gems to be removed
+      patterns = []
+      removed_deps.each do |dep|
+        patterns << /gem "#{dep.name}"/
+        patterns << /gem '#{dep.name}'/
+      end
+
+      # create a union of patterns to match any of them
+      re = Regexp.union(patterns)
+
+      new_gemfile = []
+      IO.readlines(gemfile_path).each {|line| new_gemfile << line unless line.match(re) }
+
+      # remove lone \n and append them with other strings
+      new_gemfile.each_with_index do |_line, index|
+        if new_gemfile[index + 1] == "\n"
+          new_gemfile[index] += new_gemfile[index + 1]
+          new_gemfile.delete_at(index + 1)
+        end
+      end
+
+      blocks = ["group ", "source ", "env ", "install_if"]
+      blocks.each {|block| remove_nested_blocks(new_gemfile, block) }
+
+      File.open(gemfile_path, "w") {|file| file.puts new_gemfile.join.chomp }
+    end
+
+    def remove_nested_blocks(gemfile, block_name)
+      nested_blocks = 0
+
+      # count number of nested groups
+      gemfile.each_with_index do |line, index|
+        nested_blocks += 1 if gemfile[index + 1] =~ /#{block_name}/ && line =~ /#{block_name}/
+      end
+
+      while nested_blocks >= 0
+        nested_blocks -= 1
+
+        gemfile.each_with_index do |line, index|
+          next if line !~ /#{block_name}/
+          if gemfile[index + 1] =~ /end/
```

Comment:
this should probably only match `/^\s*end\s*$` ? otherwise we could accidentally do something like remove a line with a comment that contains the work `end`